src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs

```diff
@@ -184,64 +193,104 @@ private static void AdjustSpans(ArrayBuilder<ClassifiedSpan> spans, TextSpan wid
             }
         }
 
-        private static ImmutableArray<ClassifiedSpan> MergeParts(
+        private readonly struct ClassifiedSpanIntervalIntrospector : IIntervalIntrospector<ClassifiedSpan>
+        {
+            public static readonly ClassifiedSpanIntervalIntrospector Instance = new ClassifiedSpanIntervalIntrospector();
+
+            public int GetLength(ClassifiedSpan value)
+                => value.TextSpan.Length;
+
+            public int GetStart(ClassifiedSpan value)
+                => value.TextSpan.Start;
+        }
+
+        /// <summary>
+        /// Adds all semantic parts to final parts, and adds all portions of <paramref name="syntaxParts"/> that do not
+        /// overlap with any semantic parts as well.  All final parts will be non-empty.
+        /// </summary>
+        private static void MergeParts(
```

Comment:
this algorithm is effectively brand new.  THe old one wasn't very good and definitely was not equipped to handle the case of multiple spans overlapping (for example, all the regex-classifications overlapping the syntactic classification for a string).

The new algorithm is simpler.  We just take all the semantic classifications as-is.  Then, for any syntactic classification, we see what semantic classifications is overlaps with.  And we only pick the part of the syntactic classification that doesn't intersect with the semantic classification.